### ISSUE TITLE
SMS - PubSubRendezvous not active on silo - orleans 3

### ISSUE THREAD

**Issue by stephenlautier** [dotnet/orleans#6120]
We just recently updated to netcore3 and Orleans 3 and we have some critical issue with SMS, it seems to happen mainly with the `PubSubRendezvousGrain` not activating correctly when its on "heavy load". We are calling `AddStream` on Grain Activation of certain grains, and the following is being thrown.

#### Error
![2019-11-14 17_49_48-Discover_ Odin Microservice - Errors - Kibana](https://user-images.githubusercontent.com/3908723/68877961-2f6c1780-0707-11ea-9d14-a6d284666d3c.png)

After a while, it seems to stables a bit (sometimes) - tho another error appeared, seems a silo got restarted and afterwards it couldn't locate a grain, see the following error:

> Current directory at S10.150.18.52:11111:311443752 is not stable to perform the lookup for grainId *grn/814818CD/00000000+CommonHub:crrcp1DxHhge4eUxRYc_GA (it maps to S10.150.21.226:11111:311443752, which is not a valid silo). Retry later.

![image](https://user-images.githubusercontent.com/3908723/68880851-58db7200-070c-11ea-8977-49a337cd31bb.png)

#### Using
- Docker (linux) based on `mcr.microsoft.com/dotnet/core/aspnet:3.0-buster-slim`
- K8's
- Redis as Storage Provider for SMS
- Silos Count: 9
- Server load 35-60%

Basically we had ended up reverting on prod from using Orleans 3, netcore3 for now as several services are not working correctly.

If you need anything else we be happy to help. Thanks

**Comment by sergeybykov** [comment]
I'm trying to wrap my head around this. Is this during a rolling upgrade to 3.0 from 2.x or from a fresh deployment of 3.0?

The errors you posted are supposed to be transient, caused by races when silos in the cluster go down and up. Do they not disappear after awhile?

**Comment by stephenlautier** [comment]
We had deployed and rebooted multiple times differently, but same issue remained (especially the  first error shown).

We tried
- Rebooting all cluster
- Deployed a new version (via kubes so transition from 1 version to another)
- Deployed a new version and clearing everything (via kubes so transition from 1 version to another but using a different serviceId)
- Finally, we also took down all cluster, clear redis, clear consul, start everything again

After last attempt, which we cleared everything and starting everything from fresh we were having the first issue, after ~30mins, that seemed "ok" but we started having the second error and we remained with it for about ~1hr

And initially, we had restarted due to SMS were timing out.

Just some other notes


- We use https://github.com/OrleansContrib/SignalR.Orleans, and we do get a lot of timeouts when Sending via SignalR Orleans, but essentially internal its just using SMS
- We do have high load on realtime - kafka -> signalr, and we do starting getting issues when the volumes are quite high (~20million daily atm) e.g. pro; but as stated the CPU/memory was not under high pressure. On dev/test environments we don't seem to have issues, but the volumes are much much lower.

**Comment by sergeybykov** [comment]
The first error seems to indicate that you have another stream provider (?) registered as a producer:
>*stg/72633...

SMS doesn't use system targets at all, as it uses direct messages to grains. Hence, I'm puzzled what this is from. Did you use another stream provider by chance that left a registration in a `PubSubRendezvousGrain`. It's a stretch, but I'm out of ideas here, unless you can share fuller logs.

The second error should be transient until all silos learn about the new state of the cluster, which should be very fast in 3.0.

>And initially, we had restarted due to SMS were timing out.

Do you only have grains as stream consumers or clients as well? The former case is simpler because grains lifecycle is managed. Did you see these timeouts correlated with failure to load state from Redis or any other errors?

**Comment by stephenlautier** [comment]
>Do you only have grains as stream consumers or clients as well? The former case is simpler because grains lifecycle is managed.

Yes, we do have clients as stream consumer, mainly the ones coming from SignalR Orleans https://github.com/OrleansContrib/SignalR.Orleans/blob/master/src/SignalR.Orleans/OrleansHubLifetimeManager.cs#L74

> Did you see these timeouts correlated with failure to load state from Redis or any other errors?

No Redis seemed to be stable at that point


One of the SMS timing out we have an idea why its being caused, tho **its not the same issue as to the one I had posted here**. Basically high throughput within the same stream id, which i believe they go through the same grain and it wont scale leading to timeouts.


Another error which we do get, but i highly doubt it is related is the following

##### Consumer extenions not installed error

````
Extension not installed on grain SignalR.Orleans.Clients.ClientGrain attempting to invoke type Orleans.Streams.OrleansCodeGenStreamConsumerExtensionMethodInvoker from invokable Orleans.Runtime.ActivationData
--

at Orleans.Runtime.InsideRuntimeClient.Invoke(IAddressable target, IInvokable invokable, Message message)
   at Orleans.Internal.OrleansTaskExtentions.<ToTypedTask>g__ConvertAsync|4_0[T](Task`1 asyncTask)
   at Orleans.Providers.Streams.SimpleMessageStream.SimpleMessageStreamProducerExtension.StreamConsumerExtensionCollection.DeliverToRemote(IStreamConsumerExtension remoteConsumer, StreamId streamId, GuidId subscriptionId, Object item, Boolean optimizeForImmutableData, Boolean fireAndForgetDelivery)
   at Orleans.Providers.Streams.SimpleMessageStream.SimpleMessageStreamProducer`1.OnNextAsync(T item, StreamSequenceToken token)
   at SignalR.Orleans.Core.ServerDirectoryGrain.ValidateAndCleanUp(Object serverDirectory)
   at Orleans.Runtime.GrainTimer.ForwardToAsyncCallback(Object state)
````

**Comment by sergeybykov** [comment]
@stephenlautier Have you figured this out in the long time that passed since you opened it?

**Comment by stephenlautier** [comment]
Yes, I believe so. I can close of these issues. We are on prod now with 3.x, so thats a good thing; and the last issue was also solved (not resuming correctly to a stream). Thanks!
